akka-http-core/src/test/scala/akka/testkit/ExplicitlyTriggeredScheduler.scala

```diff
@@ -0,0 +1,75 @@
+package akka.testkit
+
+import java.util.concurrent.ThreadFactory
+
+import akka.actor.{ ActorSystem, Cancellable, Scheduler }
+import akka.event.LoggingAdapter
+import com.typesafe.config.Config
+
+import scala.annotation.tailrec
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration.{ Duration, FiniteDuration }
+
+/**
+ * For testing: scheduler that does not look at the clock, but must be progressed manually by calling `timePasses`.
+ *
+ * This is not entirely realistic: jobs will be executed on the test thread instead of using the `ExecutionContext`, but does
+ * allow for faster and less timing-sensitive specs..
+ */
+class ExplicitlyTriggeredScheduler(config: Config, log: LoggingAdapter, tf: ThreadFactory) extends Scheduler {
+
+  case class Item(time: Long, interval: Option[FiniteDuration], runnable: Runnable)
+
+  var currentTime = 0L
+  var scheduled: List[Item] = Nil
+
+  override def schedule(initialDelay: FiniteDuration, interval: FiniteDuration, runnable: Runnable)(implicit executor: ExecutionContext): Cancellable =
+    schedule(initialDelay, Some(interval), runnable)
+
+  override def scheduleOnce(delay: FiniteDuration, runnable: Runnable)(implicit executor: ExecutionContext): Cancellable =
+    schedule(delay, None, runnable)
+
+  def timePasses(amount: FiniteDuration)(implicit system: ActorSystem) = {
+    executeTasks(currentTime + amount.dilated.toMillis)
```

Comment:
If the `currentTime` isn't advanced first, won't newly scheduled tasks be scheduled to an earlier time than expected? Couldn't that mean infinite loops in the worst case, e.g. if `amount = 50.ms` and one of the running tasks always schedules things to a time `< 50ms`?

Reply:
Indeed, fixed